### PR TITLE
Implement Deref for wrapper types

### LINKED ISSUE #168: More property/member consistency
Right now, there are some members that shouldn't be public, as they should be protected against mutation; e.g. `Map::width/height/tile_width`. These should be changed to properties.

### PR DESCRIPTION
**Pull request by aleokdev**
An experiment. Exposes data types but implements `Deref<Target=TData>` for wrapper types, leaving practically the same interface but also allowing the user to also use data types in more performance-critical parts.

Closes #168.

### CODE REVIEW

File: src/objects.rs [hunk 1]
```diff
@@ -28,92 +28,42 @@ pub enum ObjectShape {
...
-        self.name.as_ref()
-    }
-
-    /// The type of the object, which is arbitrary and set by the user.
-    #[inline]
-    pub fn obj_type(&self) -> &str {
-        self.obj_type.as_ref()
-    }
-
+    pub name: String,    /// The type of the object, which is arbitrary and set by the user.
```
Comment (bjorn) [inline src/objects.rs:34]: Missing newline before comment.

File: src/layers/tile/infinite.rs [hunk 2]
```diff
@@ -9,6 +9,7 @@ use crate::{
 
 use super::util::parse_data_line;
 
+/// The raw data of a [`InfiniteTileLayer`]. Does not include a reference to its parent [`Map`](crate::Map).
```
Comment (bjorn) [inline src/layers/tile/infinite.rs:12]: Does it make sense to say something about the parent here? It sounds more suitable as documentation for `TileLayerData` to me.

File: src/layers/tile/finite.rs [hunk 3]
```diff
@@ -7,6 +7,7 @@ use crate::{
 
 use super::util::parse_data_line;
 
+/// The raw data of a [`FiniteTileLayer`]. Does not include a reference to its parent [`Map`](crate::Map).
```
Comment (bjorn) [inline src/layers/tile/finite.rs:10]: Same here, the section about the parent is probably only relevant for `TileLayerData`.

**Review by bjorn — COMMENTED** [review]
Just some comments.

**Comment by bjorn** [comment]
And just to add, my first impression of this change is very good. I love that we can get rid of all those getters again!

File: src/layers/tile/finite.rs (hunk 3 above)
Comment (aleokdev) [reply src/layers/tile/finite.rs:10]: It's relevant to all map wrapped objects; they all have the same docs but with different type references

Comment (bjorn) [reply src/layers/tile/finite.rs:10]: Ah, right.

**Comment by aleokdev** [comment]
Note: Getters have stayed for members whose state has some kind of runtime guarantee (e.g. Object ID must be local within map) or  rely on the state of another object (e.g. tile layer width/height depend on the size of the tiles vector and vice-versa)

**Comment by aleokdev** [comment]
CI is broken again. Same error as when I added the SFML example, it's using some old version of the crate instead of the PR one for some reason. Sigh.

**Comment by aleokdev** [comment]
It doesn't even try to build it for some reason...

**Comment by aleokdev** [comment]
Apparently it needed the branch to be merged with master.

File: src/util.rs [hunk 4]
```diff
@@ -74,6 +74,14 @@ macro_rules! map_wrapper {
                 self.map
             }
         }
+
+        impl<'map> std::ops::Deref for $name<'map> {
+            type Target = $data_ty;
+
+            fn deref(&self) -> &'map Self::Target {
```
Comment (bjorn) [inline src/util.rs:81]: Should this thing be `#[inline]`?

File: src/layers/object.rs [hunk 5]
```diff
@@ -12,7 +12,8 @@ use crate::{
 #[derive(Debug, PartialEq, Clone)]
 pub struct ObjectLayerData {
     /// The objects present in this layer.
-    pub objects: Vec<ObjectData>,
+    // TODO: read-only property
```
Comment (bjorn) [inline src/layers/object.rs:15]: So this TODO is for a follow-up change?

File: src/objects.rs [hunk 6]
```diff
@@ -27,10 +27,8 @@ pub enum ObjectShape {
 /// Also see the [TMX docs](https://doc.mapeditor.org/en/stable/reference/tmx-map-format/#tmx-object).
 #[derive(Debug, PartialEq, Clone)]
 pub struct ObjectData {
-    /// ID of the object, which is unique per map since Tiled 0.11.
-    ///
-    /// On older versions this value is defaulted to 0.
-    pub id: u32,
+    id: u32,
+    // TODO: make public, differ somewhat from get_tile
```
Comment (bjorn) [inline src/objects.rs:31]: Why not make it public in this change? Is it more involved?

**Review by bjorn — APPROVED** [review]
Approved, but had some comments.

File: src/util.rs (hunk 4 above)
Comment (aleokdev) [reply src/util.rs:81]: Oops yeah, definitely

File: src/objects.rs (hunk 6 above)
Comment (aleokdev) [reply src/objects.rs:31]: Whether it is public or not depends on if we should check for ID uniqueness and fail if there are duplicates.

Comment (bjorn) [reply src/objects.rs:31]: I think it's fine to have `id` private, but this comment is about `tile`.

Comment (aleokdev) [reply src/objects.rs:31]: Oh. Yeah, forgot to make it public. The issue is the name, should rename it to tile_data or something.

File: src/layers/tile/mod.rs [hunk 7]
```diff
@@ -176,7 +163,7 @@ impl<'map> TileLayer<'map> {
     }
 
     /// Obtains the tile present at the position given.
-    /// 
+    ///
```
Comment (aleokdev) [inline src/layers/tile/mod.rs:166]: Unrelated change

File: src/layers/tile/finite.rs [hunk 8]
```diff
@@ -67,23 +80,11 @@ map_wrapper!(
 
 impl<'map> FiniteTileLayer<'map> {
     /// Obtains the tile present at the position given.
-    /// 
+    ///
```
Comment (aleokdev) [inline src/layers/tile/finite.rs:83]: Unrelated change

File: src/layers/tile/mod.rs [hunk 9]
```diff
@@ -17,19 +17,39 @@ pub use infinite::*;
...
+    /// Use [`LayerTile::get_tile`] if you want to obtain the [`Tile`] that this layer tile is
+    /// referencing.
+    #[inline]
+    pub fn tileset_index(&self) -> usize {
+        self.tileset_index
+    }
+
+    /// Get the layer tile's local id within its parent tileset.
+    #[inline]
+    pub fn id(&self) -> u32 {
```
Comment (aleokdev) [inline src/layers/tile/mod.rs:49]: Add guarantees

File: src/layers/object.rs (hunk 5 above)
Comment (aleokdev) [reply src/layers/object.rs:15]: No, will do later, forgot to impl it

File: src/layers/object.rs [hunk 10]
```diff
@@ -45,6 +44,13 @@ impl ObjectLayerData {
         });
         Ok((ObjectLayerData { objects, colour: c }, properties))
     }
+
+    /// Returns the data belonging to the objects contained within the layer, in no particular
+    /// order.
```
Comment (bjorn) [inline src/layers/object.rs:49]: Why would this be "in no particular order", whereas `objects()` will iterate "in the order they were declared in the TMX file"? The order is the same, right?

The order in the TMX file is relevant, btw. It is the rendering order, for objects layers with `draworder` set to `index`, and it can then be manually controlled in Tiled. But I'm not sure how much of this needs to be documented here.

**Review by bjorn — APPROVED** [review]
Approved, but might want to tweak the docs for `objects`.

**Comment by bjorn** [comment]
Pretty amazing that the build check even detects errors in commented example code.

**Comment by aleokdev** [comment]
> Pretty amazing that the build check even detects errors in commented example code.

That's because it's not a comment, it's a doctest, and it gets compiled (and executed) when `cargo test` is run.

**Review by bjorn — APPROVED** [review]

**Comment by bjorn** [comment]
> That's because it's not a comment, it's a doctest, and it gets compiled (and executed) when `cargo test` is run.

It looks like some code example in a comment though. How does it become a doctest? Are all code blocks in documentation comments doctests?

**Comment by aleokdev** [comment]
> It looks like some code example in a comment though. How does it become a doctest? Are all code blocks in documentation comments doctests?

Lines starting with `///` (or blocks starting with `/**` aren't comments; They are actually converted to `#[doc = "..."]` attributes. This is how cargo can detect docs when generating them. Code in doc attributes is compiled and executed exactly like tests. Here's more info: https://doc.rust-lang.org/rustdoc/the-doc-attribute.html
